src/pcp.rs

```diff
@@ -0,0 +1,596 @@
+    CircuitInLen,
+
+    /// This error is returned by `collect` if the input slice is empty.
+    #[error("collect requires at least one input")]
+    CollectInLen,
+
+    /// This error is returned by `collect` if the two or more verifier shares have different
+    /// gadget arities.
+    #[error("collect inputs have mismatched gadget arity")]
+    CollectGadgetInLenMismatch,
+
+    /// Returned if an FFT operation propagates an error.
+    #[error("FFT error")]
+    Fft(#[from] FftError),
+
+    /// When evaluating a gadget on polynomials, this error is returned if the input polynomials
+    /// don't all have the same length.
+    #[error("gadget called on polynomials with different lengths")]
+    GadgetPolyInLen,
+
+    /// When evaluating a gadget on polynomials, this error is returned if the slice allocated for
+    /// the output polynomial is too small.
+    #[error("sliice allocated for gadget output is too small")]
+    GadgetPolyOutLen,
+
+    /// The proof string is either too large or too short for the given type.
+    #[error("attempted query on proof with invalid length")]
+    QueryProofLen,
+}
+
+/// A data type. Implementations of this trait specify an arithmetic circuit that determines
+/// whether a given value is valid.
+pub trait Type<F, G>: Sized
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    /// Evalauates the validity circuit on the given input (i.e., `self`) and returns the output.
+    /// Slice `rand` is the random input consumed by the validity circuit.
+    ///
+    /// ```
+    /// use prio::pcp::types::Boolean;
+    /// use prio::pcp::Type;
+    /// use prio::field::{FieldElement, Field64};
+    ///
+    /// type F = Field64;
+    /// type T = Boolean<F>;
+    ///
+    /// let x = T::new(false);
+    ///
+    /// let mut rand: Vec<F> = Vec::with_capacity(x.valid_rand_len());
+    /// for _ in 0..rand.len() {
+    ///     rand.push(F::rand());
+    /// }
+    ///
+    /// let v = x.valid(&mut x.gadget(0), &rand).unwrap();
+    /// assert_eq!(v, F::zero());
+    /// ```
+    fn valid(&self, g: &mut dyn GadgetCallOnly<F>, rand: &[F]) -> Result<F, PcpError>;
+
+    /// Returns a reference to the underlying data.
+    fn as_slice(&self) -> &[F];
+
+    /// Constructs an instance of this type from the given data without checking whether the data
+    /// is valid. This method takes ownership of `data`.
+    fn clone_with(&self, data: Vec<F>) -> Self;
+
+    /// The length of the random input expected by the validity circuit.
+    fn valid_rand_len(&self) -> usize;
+
+    /// The number of calls to the gadget made when evaluating the validity circuit.
+    fn valid_gadget_calls(&self) -> usize;
+
+    /// Returns an instance of the gadget associated with the validity circuit. `in_len` is the
+    /// maximum degree of the each polynomial input to `call_poly`. If this function is never used
+    /// by the caller, then it is save to set `in_len == 0`.
```

Comment:
(I took the second suggestion but not the first. `call_poly` takes in a sequence of polynomials. `in_len` is the length of each of these polynomials.)

"this function " = `Gadget::call_poly`. (Corrected the text to clarify.)

Maybe it would be less confusing if the method signature were this?
```rust
fn gadget(&self, in_len: Option<usize>) -> G;
```

Incidentally, the `in_len` is helpful because it allows us to pre-compute some state required by FFT operations (See `n` and  `n_inv` in `struuct PolyEval` in `gadget.rs` in the next PR.)

Reply:
There's one more typo in the comment I flagged but otherwise this makes sense to me now.